omegaconf/dictconfig.py

```diff
@@ -243,6 +230,24 @@ def is_typed(c: Any) -> bool:
             )
             raise ValidationError(msg)
 
+    def _check_assign_with_wrap_value(self, key: Any, value: Any) -> None:
+        from omegaconf.omegaconf import _maybe_wrap
+
+        element_type = self._metadata.element_type
+        dummy_value = copy.deepcopy(value)
+        optional = dummy_value._metadata.optional
+        dummy_parent = DictConfig(content={})
+        _maybe_wrap(
+            ref_type=element_type,
+            key=key,
+            value=dummy_value._value(),
+            is_optional=optional,
+            parent=dummy_parent,
+        )
```

Comment:
It looks like you are using maybe_wrap to do the validation and you are not doing anything with the return value.
It seems a wasteful, If this should really get wrapped, it will happen twice.

Also, what's the deal with the dummy parent and you using it's value in the strangest and probably incorrect way?

Reply:
What happened with the assignment of `User(age=1, name="foo")` is that when getting its data transformed everything to value nodes so, when you wrapped a value node it didn't validate it since it didn't really wrap a new node so I wrapped a new value with the node's value.